src/Features/Core/Portable/ForeachToFor/AbstractForeachToForCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,346 @@
+            }
+
+            var model = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            var semanticFact = document.GetLanguageService<ISemanticFactsService>();
+            var foreachInfo = GetForeachInfo(semanticFact, model, foreachStatement, cancellationToken);
+            if (foreachInfo == null)
+            {
+                return;
+            }
+
+            context.RegisterRefactoring(
+                new ForeachToForCodeAction(
+                    FeaturesResources.Convert_foreach_to_for,
+                    c => ConvertForeachToForAsync(document, foreachInfo, c)));
+        }
+
+        protected string CreateUniqueMethodName(SemanticModel model, SyntaxNode contextNode, string baseName)
+        {
+            Contract.ThrowIfNull(contextNode);
+            Contract.ThrowIfNull(baseName);
+
+            return NameGenerator.GenerateUniqueName(baseName, string.Empty,
+               name => model.LookupSymbols(contextNode.SpanStart, /*container*/null, name).Length == 0);
+        }
+
+        protected T AddElasticAnnotation<T>(T node, SyntaxTrivia elasticTrivia) where T : SyntaxNode
+        {
+            var first = node.GetFirstToken();
+            node = node.ReplaceToken(first, first.WithPrependedLeadingTrivia(elasticTrivia));
+
+            var last = node.GetLastToken();
+            return node.ReplaceToken(last, last.WithAppendedTrailingTrivia(elasticTrivia));
+        }
+
+        protected T AddRenameAnnotation<T>(T node, string name) where T : SyntaxNode
+        {
+            var token = node.DescendantTokens().FirstOrDefault(t => t.Text == name);
+            if (token == default(SyntaxToken))
+            {
+                return node;
+            }
+
+            return node.ReplaceToken(token, token.WithAdditionalAnnotations(RenameAnnotation.Create()));
+        }
+
+        private ForeachInfo GetForeachInfo(ISemanticFactsService semanticFact, SemanticModel model, SyntaxNode foreachStatement, CancellationToken cancellationToken)
+        {
+            var operation = model.GetOperation(foreachStatement, cancellationToken) as IForEachLoopOperation;
+            if (operation == null || operation.Locals.Length != 1)
+            {
+                return null;
+            }
+
+            var foreachVariable = operation.Locals[0];
+            if (foreachVariable == null)
+            {
+                return null;
+            }
+
+            if (!CheckForeachVariable(model, foreachVariable, foreachStatement))
+            {
+                return null;
+            }
+
+            var foreachCollection = RemoveImplicitConversion(operation.Collection);
+            if (foreachCollection == null)
+            {
+                return null;
+            }
+
+            string explicitCast;
+            string countName;
+            GetInterfaceInfo(semanticFact, model, foreachVariable, foreachCollection, out explicitCast, out countName);
```

Comment:
why not out var?

Reply:
sure :) I just used extract method. I guess we need to update extract method to use new language features.